### DISCUSSION TITLE
How to handle dependencies already installed?

### DISCUSSION THREAD

**Discussion by pjljvandelaar** [alire-project/alire#994] [Q&A]
Dear Alire developers,

Many of the users of my alire crate have GNAT Pro and LibAdaLang already installed on their PC.
Building my crate takes quite some time, since the installed version of GNATColl and LibAdaLang are not reused.
In fact, we have regularly problems due to the usage of different versions of libraries in different steps of the build, compile, and release sequence. See e.g. https://github.com/alire-project/alire/issues/982 and executions that state `the process exited with status -1073741819`.

Users of alire can choose that the gnat version and/or the gprbuild version installed in the environment are used.
Should we also allow users to choose that the GNATColl version and LibAdaLang version installed in the environment are used?
Or could alire detect the dependencies are available in the environement and then use them?

I am curious to your answers!

Greetings,

Pierre

**Comment by Fabien-Chouteau** [comment]
Hi @pjljvandelaar,

GNAT pro and Alire workflows are not always compatible.

One option here would be to remove `libadalang` from the list of dependencies in you crate, this way gprbuild will use the one from GNAT Pro.

> Users of alire can choose that the gnat version and/or the gprbuild version installed in the environment are used.
> Should we also allow users to choose that the GNATColl version and LibAdaLang version installed in the environment are used?
> Or could alire detect the dependencies are available in the environement and then use them?

This is not planed at the moment and I am not even sure if it is doable. I don't think there is a reliable way to get the version of a library in the system or in GNAT Pro install.

Alire is a source package manager so it is part of the workflow to build from sources. It has drawbacks and advantages.

**Reply by pjljvandelaar** [reply]
Your suggestion 
`One option here would be to remove libadalang from the list of dependencies in you crate`
has as disadvantage that users that don't have libadalang get errors and might not know which crate they are missing.

**Reply by Fabien-Chouteau** [reply]
Indeed it's a trade-off.

**Comment by mosteo** [comment]
Hi @pjljvandelaar. Another possibility worth a try, if the installed versions have project files with the same name as the alire ones, it's to instruct your users to `alr pin` to them. This will override the regular dependency.

If project file names do change, you could prepare a proxy project that withed the installed versions, and pin to the proxy project (which should have the proper project names).

I haven't tested this particular case, so I'd be interested in knowing if you make it work.

Although, removing the dependencies and setting an environment variable as Fabien suggested is probably less work...

**Reply by pjljvandelaar** [reply]
Nice suggestion. I will experiment with it!

**Reply by pjljvandelaar** [reply]
Am I missing something: neither
[[pins]]
libadalang = { path = "C:/GNATPRO/22.1/include/libadalang" }
nor
[[pins]]
libadalang = { path = "C:/GNATPRO/22.1/share/gpr" }
works.

And I can't combine them:
`invalid path name "C:\GNATPRO\22.1\include\libadalang;C:\GNATPRO\22.1\share\gpr"`

**Reply by pjljvandelaar** [reply]
How to specify the dependencies of libadalang?

In the first case I get
```
rejuvenation-utils.adb:4:06: error: file "gnatcoll.ads" not found
rejuvenation-utils.adb:4:33: error: "GNATCOLL" is undefined
rejuvenation-utils.adb:5:06: error: file "langkit_support.ads" not found
rejuvenation-utils.adb:5:33: error: "Langkit_Support" is undefined
```
so here the dependencies of libadalang are the problem.

While in the latter case I get
```
rejuvenation.ads:3:06: error: file "libadalang.ads" not found
rejuvenation.ads:3:37: error: "Libadalang" is undefined (more references follow)
rejuvenation.ads:4:06: error: file "libadalang.ads" not found
rejuvenation.ads:8:61: error: "Ada_Node_Kind_Type" is undefined
```
So here libadalang is the problem!

Order of files is causing the difference in the first lines: rejuvenation.ads always has these errors.

**Reply by pjljvandelaar** [reply]
I seem to be able to add multiple paths to a pin
using e.g.
```
libadalang = { path = "C:/GNATPRO/22.1/include/libadalang" }
gnatcoll = { path = "C:/GNATPRO/22.1/include/gnatcoll" }
gnatcoll_gmp = { path = "C:/GNATPRO/22.1/include/gnatcoll_gmp" }
gnatcoll_iconv = { path = "C:/GNATPRO/22.1/include/gnatcoll_iconv" }
langkit_support = { path = "C:/GNATPRO/22.1/include/langkit_support" }
xmlada = { path = "C:/GNATPRO/22.1/include/xmlada" }
```
but it still doesn't find the dependencies...

**Reply by mosteo** [reply]
For this to work you'll need to pin directly the folder containing the project files of each dependency.

**Reply by pjljvandelaar** [reply]
I wil experiment some more with "C:\GNATPRO\22.1\share\gpr" since
that contains the project files of the dependencies, right?

![image](https://user-images.githubusercontent.com/18348654/165716748-3975c4a4-6877-4ba6-a472-8222962b99c0.png)

**Reply by mosteo** [reply]
Right, and the pin system should detect that as a "raw" project (no manifest, only the project file)

**Reply by pjljvandelaar** [reply]
Despite that libadalang.gpr contains
`for Source_Dirs use ("../../include\libadalang\");`
I still get
`rewriters_sequence.ads:2:06: error: file "libadalang.ads" not found`

Has it something to do with Primary_Source_Dirs? See https://gt3-prod-1.adacore.com/#/tickets/V428-008

**Reply by mosteo** [reply]
I don't have access to that ticket, sorry.

I'm not sure if there's some way to see the final search paths in use by gprbuild. That could help to ascertain what's happening.
